### ISSUE TITLE
Binaries and config distribution 

### ISSUE THREAD

**Issue by Jabster28** [haveno-dex/haveno#931]
Before v1.0.3 came out, I made a [branch on my fork](https://github.com/Jabster28/haveno/tree/flatpak) for adding Flatpak binaries, since they have a lot of safety features like sandboxing and filesystem access. However, now that Haveno's going for a more decentralized approach with separate networks, I'm unsure of how we should deal with distributing the Flatpaks (and other binaries).

Flatpak software managers will let you install `.flatpak`s if you open them, but the way they're intended to work is similar to most other package managers. They expect to have a repository hosted somewhere that contains the files and update data, so that they can update automatically in the background and do any other tasks that need doing. The same can be said for the `.deb`s and `.rpm`s that our CI builds. But the decision of what repos and what to upload to which repos depends on the way we decide to implement the network configs.

Whatever choice we make in sharing binaries should be kept consistent, in my opinion. I'm building on the ideas from #911 and #903 in their thoughts on ways implement to this, and I feel like this decision shouldn't be taken lightly. I think the different ways of sharing them come down to these two questions:

- [ ] Should the "network configuration" (arb keys, seed nodes, fees, appdir, etc.) be configurable at runtime?
	- This is currently not possible, and as such all Haveno networks must build their own binaries and distribute them themselves if they want to have compiled artifacts. This could be confusing for some, as downloading an exe from the "official" Haveno repository will simply not work.
	- If this is the case, virtually all Haveno users could use this upstream repositories (or anywhere else) for their app downloads, and simply download the config from the network(s) they choose. This would also make verification and scam checks easier, as all the binaries should have the same checksum.
- [ ] Should a "network configuration" be configurable at compile time?
	- If configs can be set at compile time, and can't be overridden, we end up with a similar situation with how it is now; users can opt to get their binaries from their specific network via whatever means the network offers. However, they can also download the "official" Haveno binaries and download a JSON/TOML/whatever to use their network.
	- Allowing a "compile-time config" to be overridden by a runtime config would result in a bad UX in my opinion. A user could potentially try to check the integrity of what they think is an installation of network A (since, to them, it runs on network A) while the compile-time config inside the binary defines it as a network B installation, resulting in a failed integrity check.
	- Disabling compile-time configs would mean that every binary running a non-forked version of Haveno could be checked easily, and you would never need to re-install the application to change networks. This would, however, force users who arrive at a network to download two files - the official build and a network config.
_The second bullet point assumes the first in these examples._

From the (little) thought I've put into this, I think allowing a non-overridable (but still optional) compile-time configuration - while also allowing for a file to be read if there's no config present - is the best way to deal with configs. This would allow for the Haveno group to distribute binaries and upload them to Flathub and other centralised software repositories and making integrity checks easy, and it would also allow for networks to set up their own Debian PPAs/Flatpak repos for a more trustless design.

## Appendix

"Compile-time" and "Runtime" are used in this to refer to whether a specific config is included into the binaries when compiled or not. By compile-time, I mean the time when the resulting binaries are built (and as such the contents of such binaries). Runtime implies that the configuration can change while the application stays installed by, e.g., moving a file, restoring a different backup, resetting the config, etc.

cc @woodser @chefnaphtha @mcneb10 for their thoughts

**Comment by shortwavesurfer2009** [comment]
I see one potential issue and that would be the management of the wallet files. Say I am a user of Alice Network, and I decide to switch to Bob Network. When I make that change, at least right now, the folder would change its name from "Haveno-alice to " Haveno-bob". This, at least as of now, would cause me to lose all of my settings, such as dark mode, my accounts that I've created for fiat and crypto, and my wallet with its balance. I still think this could be done, but I would add a big warning to the process. And if the user doesn't read it, it's on them, not us.

**Comment by chefnaphtha** [comment]
@shortwavesurfer2009 agreed, i think making the non-network-specific configs global is a good idea

otherwise, yeah, i don't like having the only option of network configuration be at compile time. the official testnet should be the default network, hardcoded as a fallback if no network file is specified

**Comment by Jabster28** [comment]
> @chefnaphtha i also don't think anyone would do integrity checks like that, what would they compare to?
Yeah, you may be right actually. Now that I think about it, someone who can understand how to switch networks would probably know not to check the sums until they update the actual application. I'll edit that bit later.

**Comment by Jabster28** [comment]
@shortwavesurfer2009 Yeah, your example does sound like a problem. Perhaps instead of editing a file, the "proper" way to switch networks should be to select the file using a menu in the interface? That way the app could do some things automatically like throw up some warnings, copy the wallet files, archive the previous orders etc.

If we're doing it that way, I guess editing the file manually should be advised against.

**Comment by woodser** [comment]
I think it should be built with the network config hardcoded for simplicity, just like the current installers. Maybe in the future Haveno will support inputting custom network config e.g. on startup.

**Comment by shortwavesurfer2009** [comment]
Isnt that what #1073 aims to do?

**Comment by woodser** [comment]
Yeah #1073 aims to support configurable network config, which should be out of scope for Flatpak and AppImage installers.

**Comment by shortwavesurfer2009** [comment]
But a flatpack/appimage should be of the base clirnt and allow you as a user to connect to whatever network they want. Otherwise you have haveno-dex appimage, haveno-reto appimage, haveno-network3 appimage, etc for however many networks exist. There really should be one client downloaded from here that everyone can ise. I see people not realizing that deb and rs exist for reto and are confused as to why cloning and compiling haveno-dex repo doesnt get them on reto

**Comment by woodser** [comment]
> a flatpack/appimage should be of the base client and allow you as a user to connect to whatever network they want.

The Flatpak / AppImage should be the same as the other installers by faithfully representing the capabilities of the repo it's built from. Then if Haveno supports customizing the network config in the future, all installers inherit this capability just the same. We would not want to introduce custom capabilities specifically to the Flatpak / AppImage installers and maintain that capability only for those installers.

**Comment by shortwavesurfer2009** [comment]
Oh, okay, i understand now
